### PR TITLE
 Add jib:buildTar mojo to maven plugin

### LINKED ISSUE #514: Allow user to build image as tar file that can be 'docker load'ed
https://github.com/GoogleContainerTools/jib/blob/master/proposals/build_tarball.md

- [x] Add private output path configuration parameter
- [x] Rename `BuildTarballAndLoadDockerStep` to `LoadDockerStep`
- [x] Add `WriteTarFileStep`
- [x] Add `BuildSteps#forBuildToTarball()`
- [x] Add new task to Gradle plugin
- [x] Add new mojo to Maven plugin

### PR DESCRIPTION
**Pull request by TadCordle**
After #567
Fixes #514

### CODE REVIEW

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java [hunk 1]
```diff
@@ -64,17 +63,18 @@ public void execute() throws MojoExecutionException {
     }
 
     // Parses 'from' and 'to' into image reference.
+    MavenProjectProperties mavenProjectProperties =
+        MavenProjectProperties.getForProject(getProject(), mavenBuildLogger);
     ImageReference baseImage = parseImageReference(getBaseImage(), "from");
-    ImageReference targetImage = getDockerTag(mavenBuildLogger);
+    ImageReference targetImage =
+        mavenProjectProperties.getDockerTag(getTargetImage(), mavenBuildLogger);
```
Comment (briandealwis) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildDockerMojo.java]: `getDockerTag()` -> `getTargetDockerTag()`?

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    RegistryClient.setUserAgentSuffix(USER_AGENT_SUFFIX);
+
+    try {
+      BuildStepsRunner.forBuildTar(
+              Paths.get(getProject().getBuild().getDirectory()).resolve("jib.tar"),
+              buildConfiguration,
+              mavenProjectProperties.getSourceFilesConfiguration())
+          .build(HELPFUL_SUGGESTIONS);
+      getLog().info("");
```
Comment (briandealwis) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java]: It feels like we should be calling `MavenProjectHelper#attachArtifact()` here?

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java [hunk 3]
```diff
@@ -142,4 +144,30 @@ String getMainClass(JibPluginConfiguration jibPluginConfiguration) throws MojoEx
...
+  /**
+   * Returns an {@link ImageReference} parsed from the configured target image, or one of the form
+   * {@code project-name:project-version} if target image is not configured
+   *
+   * @param targetImage the configured target image reference (can be empty)
+   * @param mavenBuildLogger the logger used to notify users of the target image parameter
+   * @return an {@link ImageReference} parsed from the configured target image, or one of the form
+   *     {@code project-name:project-version} if target image is not configured
+   */
+  ImageReference getDockerTag(@Nullable String targetImage, MavenBuildLogger mavenBuildLogger) {
```
Comment (briandealwis) [inline jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java]: This name should indicate that it's generating a target image tag.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java (hunk 2 above)
Comment (coollog) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/BuildTarMojo.java]: I'm not too familiar with this, but I don't think we'd want to be attaching it as an artifact. I think if a user really wanted it as part of their build artifacts, they can use something like the [build-helper-maven-plugin](https://www.mojohaus.org/build-helper-maven-plugin/attach-artifact-mojo.html).

**Pull request merged**